[CS] Debug: Print ASTNode for contextual type

hamishknight opened the pull request:
If we don't have an associated TypeRepr for a contextual type in the constraint system, dump the ASTNode that it's for instead.

Before:

```
Contextual Type: $T2
```

After:

```
Contextual Type: $T2 at DeclRef@/Users/hamish/src/swift-test-arena/swift-test-arena/main.swift:1493:21
```

hamishknight commented:
@swift-ci please smoke test

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -7159,3 +7159,20 @@ ASTNode constraints::findAsyncNode(ClosureExpr *closure) {
+    if (SM) {
+      out << '@';
+      pattern->getLoc().print(out, *SM);
+    }
```
Could you please leave a `TODO(diagnostics)` to implement the rest?

xedin replied:
Thank you!

hamishknight committed:
[CS] Debug: Print ASTNode for contextual type

If we don't have an associated TypeRepr for a
contextual type in the constraint system, dump the
ASTNode that it's for instead.

Before:

```
Contextual Type: $T2
```

After:

```
Contextual Type: $T2 at DeclRef@/Users/hamish/src/swift-test-arena/swift-test-arena/main.swift:1493:21
```

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test Linux
